drivers/input/sbutton.c

```diff
@@ -0,0 +1,269 @@
+#define DEV_NAMELEN     11
+
+#define KEY_PRESS       true
+#define KEY_RELEASE     false
+
+#define KEY_THRESH_MS   500
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct sbutton_dev_s
+{
+  FAR const struct sbutton_config_s *config;  /* Board configuration data */
+
+  mutex_t lock;         /* Exclusive access to dev */
+  clock_t start;        /* Clock tick when the key was pressed */
+  clock_t end;          /* Clock tick when the key was released */
+  bool    pressed;      /* Keep previous status of button */
+  struct  work_s work;  /* Supports the interrupt handling "bottom half" */
+
+  /* Keyboard lowerhalf of the registered keyboard */
+
+  struct keyboard_lowerhalf_s lower;
+};
+
+/****************************************************************************
+ * Static Function Prototypes
+ ****************************************************************************/
+
+static int sbutton_interrupt(int irq, FAR void *context, FAR void *arg);
+static void sbutton_worker(FAR void *arg);
+static void sbutton_putbuffer(FAR struct sbutton_dev_s *priv,
+                               uint8_t keycode);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sbutton_worker
+ ****************************************************************************/
+
+static void sbutton_worker(FAR void *arg)
+{
+  FAR struct sbutton_dev_s *priv = (FAR struct sbutton_dev_s *)arg;
+  uint8_t                   state;
+  int                       ret;
+
+  ret = nxmutex_lock(&priv->lock);
+  if (ret < 0)
+    {
+      return;
+    }
+
+  /* Read the status of the button */
+
+  state = priv->config->pin_status();
+
+  /* If the user just pressed the button, start measuring */
+
+  if (state == KEY_PRESS && !priv->pressed)
+    {
+      iinfo("Button pressed\n");
+
+      priv->pressed = true;
+      priv->start   = clock_systime_ticks();
+    }
+  else
+    {
+      if (state == KEY_RELEASE && priv->pressed)
+        {
+          uint32_t elapsed;
+
+          iinfo("Button released\n");
+
+          priv->pressed = false;
+          priv->end = clock_systime_ticks();
+
+          elapsed = priv->end - priv->start;
+          if (elapsed < MSEC2TICK(KEY_THRESH_MS))
+            {
+              iinfo("KEY_1\n");
+              keyboard_event(&priv->lower, ASCII_A, KEYBOARD_PRESS);
+            }
+          else
+            {
+              iinfo("KEY_2\n");
+              keyboard_event(&priv->lower, ASCII_B, KEYBOARD_PRESS);
```

Comment:
Maybe it would be good to let end user decide which key to trigger with long and short presses? Best would be to let user decide to use any keyboard keycode. If not that, at least let him swap behavior.

Reply:
Ok, I will make it to become a Kconfig option, so the user will pass the ASCII code of the key